packages/roosterjs-content-model-core/lib/coreApi/addUndoSnapshot/addUndoSnapshot.ts

```diff
@@ -11,19 +17,54 @@ import type { AddUndoSnapshot, Snapshot } from 'roosterjs-content-model-types';
  * when undo/redo to this snapshot
  */
 export const addUndoSnapshot: AddUndoSnapshot = (core, canUndoByBackspace, entityStates) => {
-    const { lifecycle, physicalRoot, undo } = core;
+    const { lifecycle, physicalRoot, logicalRoot, undo } = core;
     let snapshot: Snapshot | null = null;
 
     if (!lifecycle.shadowEditFragment) {
         // Need to create snapshot selection before retrieve innerHTML since HTML can be changed during creating selection when normalize table
         const selection = createSnapshotSelection(core);
         const html = physicalRoot.innerHTML;
 
+        // give plugins the chance to share entity states to include in the snapshot if the logical root is in an entity
+        const entityWrapper = findClosestEntityWrapper(logicalRoot, core.domHelper);
+        if (!entityStates && entityWrapper) {
+            const entityFormat = parseEntityFormat(entityWrapper);
+            if (entityFormat.entityType && entityFormat.id) {
+                const event = <EntityOperationEvent>{
+                    eventType: 'entityOperation',
+                    operation: 'snapshotEntityState',
+                    entity: {
+                        type: entityFormat.entityType,
+                        id: entityFormat.id,
+                        wrapper: entityWrapper,
+                        isReadonly: entityFormat.isReadonly,
+                    },
+                    state: undefined,
+                };
+
+                core.api.triggerEvent(core, event, false);
+
+                // copy out any entity states from the plugins
+                if (event.state) {
+                    entityStates = [
+                        {
+                            type: entityFormat.entityType,
+                            id: entityFormat.id,
+                            state: event.state,
+                        },
+                    ];
+                }
+            }
+        }
+
+        const logicalRootPath =
+            logicalRoot === physicalRoot ? [0] : getPath(logicalRoot, 0, physicalRoot);
```

Comment:
Are you sure it will be "[0]"? I feel it should be empty array. Can you double check?

Reply:
Yes, I checked and it is `[0]`

Reply:
I feel there should be a code bug in existing code. When child and parent node are the same, we should return [] for selection path. 

I think it is because this is not a normal case, so we have never seen it cause issue before. But anyway, let's just skip the path here if logical root is physical root to make sure when restore, the correct root can be set.

Reply:
Ok, I just verified the selection path is correct. But still, let's do not set selection path for logical root if it is the same with physical root since we don't need to.